contracts/contracts/token/OUSD.sol

```diff
@@ -511,93 +491,181 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
+        _adjustGlobals(-oldCredits.toInt256(), balance.toInt256());
 
-        // Mark explicitly opted out of rebasing
-        rebaseState[msg.sender] = RebaseOptions.OptOut;
-        emit AccountRebasingDisabled(msg.sender);
+        emit AccountRebasingDisabled(_account);
     }
 
     /**
      * @dev Modify the supply without minting new tokens. This uses a change in
      *      the exchange rate between "credits" and OUSD tokens to change balances.
      * @param _newTotalSupply New total supply of OUSD.
      */
-    function changeSupply(uint256 _newTotalSupply)
-        external
-        onlyVault
-        nonReentrant
-    {
-        require(_totalSupply > 0, "Cannot increase 0 supply");
+    function changeSupply(uint256 _newTotalSupply) external onlyVault {
+        require(totalSupply > 0, "Cannot increase 0 supply");
 
-        if (_totalSupply == _newTotalSupply) {
+        if (totalSupply == _newTotalSupply) {
             emit TotalSupplyUpdatedHighres(
-                _totalSupply,
+                totalSupply,
                 _rebasingCredits,
                 _rebasingCreditsPerToken
             );
             return;
         }
 
-        _totalSupply = _newTotalSupply > MAX_SUPPLY
+        totalSupply = _newTotalSupply > MAX_SUPPLY
             ? MAX_SUPPLY
             : _newTotalSupply;
 
-        _rebasingCreditsPerToken = _rebasingCredits.divPrecisely(
-            _totalSupply.sub(nonRebasingSupply)
-        );
+        _rebasingCreditsPerToken =
+            (_rebasingCredits * 1e18) /
+            (totalSupply - nonRebasingSupply);
 
         require(_rebasingCreditsPerToken > 0, "Invalid change in supply");
 
-        _totalSupply = _rebasingCredits
-            .divPrecisely(_rebasingCreditsPerToken)
-            .add(nonRebasingSupply);
+        totalSupply =
+            ((_rebasingCredits * 1e18) / _rebasingCreditsPerToken) +
+            nonRebasingSupply;
 
         emit TotalSupplyUpdatedHighres(
-            _totalSupply,
+            totalSupply,
             _rebasingCredits,
             _rebasingCreditsPerToken
         );
     }
+
+    function delegateYield(address from, address to) external onlyGovernor {
+        require(from != address(0), "Zero from address not allowed");
+        require(to != address(0), "Zero to address not allowed");
+
+        require(from != to, "Cannot delegate to self");
+        require(
+            yieldFrom[to] == address(0) &&
+                yieldTo[to] == address(0) &&
+                yieldFrom[from] == address(0) &&
+                yieldTo[from] == address(0),
+            "Blocked by existing yield delegation"
+        );
+        RebaseOptions stateFrom = rebaseState[from];
+        RebaseOptions stateTo = rebaseState[to];
+
+        require(
+            stateFrom == RebaseOptions.NotSet ||
+                stateFrom == RebaseOptions.StdNonRebasing ||
+                stateFrom == RebaseOptions.StdRebasing,
```

Comment:
I'm wondering if we should remove `stateFrom == RebaseOptions.StdRebasing` from  the allowed entry states for from. The reasoning is that this prevents governance from taking over yield from someone one by force. 

In the current code governance could take over yield from an EOA, or from a contract that had opted in to yield. If we did not have this, a contract that had marked itself as accepting yield would be safe, and any EOA could at least opt-in (from NonSet) during the timelock if they wanted to prevent their being taken.

Reply:
Makes me think of a couple of things: 
1. What you point out Daniel. That a contract could explicitly call rebaseOptIn to protect itself from having the yield "stolen" by a runaway (?) governance
2. Some pool contract could have the code in place to call `rebaseOptIn` on OUSD contract and after that it decides to delegate its yield. Now it needs a contract upgrade to be able to call the `rebaseOptOut` in order for the governance transaction to pass
3. A pool contract might decide to delegate the yield to another contract and then by calling `rebaseOptIn` it could intentionally block to delegate the yield. 

I think all 3 above can be conflated to "who should have the privilege/power?": 
- should any account have the agency to be able to block a governance yield delegation call
- should xOGN governance be able to override any user decision

To me it makes sense to leave the power in the hands of xOGN. Since it already can replace OUSD contract implementation and deploy malicious code that can drain user's complete balances not only the yield.

Reply:
Yes, I'm wanting restrictions on what governance can easily do. It's not just about a battle between two sides, but also prevents mistakes. If a contract opt's in, then it's has been specifically coded to use yield, and might even be using the creditsBalances to track do math with. Changing the behavior from the outside could be bad.

Theoretically, governance could upgrade the contracts to steal all the money, but that doesn't mean we should have a steal all the money function in the code. :D